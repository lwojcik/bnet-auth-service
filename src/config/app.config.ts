import { registerAs } from '@nestjs/config';
import { trueStringToBoolean } from '../utils/trueStringToBoolean';
import { APP, APP_INFO, DEFAULTS } from '../common/constants';

const { env } = process;
const defaultValue = DEFAULTS.app;

export const appConfig = registerAs('app', () => ({
  name: APP_INFO.name,
  environment: env[APP.environment],
  host: env[APP.host] || defaultValue.host,
  port: parseInt(env[APP.port], 10) || defaultValue.port,
  enableCors:
    trueStringToBoolean({ value: env[APP.enableCors] }) ||
    defaultValue.enableCors,
  corsOrigin: env[APP.corsOrigin] || defaultValue.corsOrigin,
}));
