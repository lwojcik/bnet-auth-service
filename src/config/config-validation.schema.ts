import * as Joi from 'joi';
import { CronExpression } from '@nestjs/schedule';
import { validateRegionName } from '../common/validators/validateRegionName.validator';

const appSchema = {
  NODE_ENV: Joi.string().default('production'),
  BAS_APP_HOST: Joi.string().default('0.0.0.0'),
  BAS_APP_PORT: Joi.string().default(3000),
  BAS_APP_CORS_ORIGIN: Joi.optional(),
};

const redisSchema = {
  BAS_REDIS_HOST: Joi.string().optional().default('redis'),
  BAS_REDIS_PORT: Joi.string().optional().default('6379'),
  BAS_REDIS_PASSWORD: Joi.any().optional(),
  BAS_REDIS_TTL_SECS: Joi.string().optional().default('2000'),
  BAS_REDIS_DB: Joi.string().optional().default('0'),
};

const battlenetSchema = {
  BAS_BATTLENET_REGION: Joi.string().required().custom(validateRegionName),
  BAS_BATTLENET_CLIENT_ID: Joi.string().required(),
  BAS_BATTLENET_CLIENT_SECRET: Joi.string().required(),
};

const throttleSchema = {
  BAS_THROTTLE_LIMIT: Joi.string().optional().default('300'),
  BAS_THROTTLE_TTL_SECS: Joi.string().optional().default('60'),
};

const authSchema = {
  BAS_AUTH_USERNAME: Joi.string().optional(),
  BAS_AUTH_JWT_SECRET: Joi.string().optional(),
};

const httpsSchema = {
  BAS_HTTPS_KEY_PATH: Joi.string().optional(),
  BAS_HTTPS_CERT_PATH: Joi.string().optional(),
};

const cronSchema = {
  BAS_CRON_PATTERN: Joi.string()
    .optional()
    .default(CronExpression.EVERY_30_MINUTES),
};

export const configValidationSchema = Joi.object({
  ...appSchema,
  ...redisSchema,
  ...battlenetSchema,
  ...throttleSchema,
  ...authSchema,
  ...httpsSchema,
  ...cronSchema,
});
