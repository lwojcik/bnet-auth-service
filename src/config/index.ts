export * from './app.config';
export * from './auth.config';
export * from './battlenet.config';
export * from './endpoints.config';
export * from './redis.config';
export * from './throttle.config';
