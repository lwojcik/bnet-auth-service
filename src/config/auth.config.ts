import { registerAs } from '@nestjs/config';
import { trueStringToBoolean } from '../utils/trueStringToBoolean';
import { AUTH, DEFAULTS } from '../common/constants';

const { env } = process;
const defaultValue = DEFAULTS.auth;

export const authConfig = registerAs('auth', () => ({
  enable:
    trueStringToBoolean({ value: process.env[AUTH.enable] }) ||
    defaultValue.enable,
  username: env[AUTH.username] || defaultValue.username,
  jwtSecret: env[AUTH.jwtSecret] || defaultValue.jwtSecret,
}));
