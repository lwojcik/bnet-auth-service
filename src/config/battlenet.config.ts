import { registerAs } from '@nestjs/config';
import { BATTLENET } from '../common/constants';

const { env } = process;

export const battleNetConfig = registerAs('battlenet', () => ({
  region: env[BATTLENET.region],
  clientId: env[BATTLENET.clientId],
  clientSecret: env[BATTLENET.clientSecret],
}));
