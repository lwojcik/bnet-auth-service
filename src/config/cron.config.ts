import { registerAs } from '@nestjs/config';

export const cronConfig = registerAs('cron', () => ({
  enable: process.env.BAS_CRON_ENABLE === 'true',
  pattern: process.env.BAS_CRON_PATTERN,
}));
