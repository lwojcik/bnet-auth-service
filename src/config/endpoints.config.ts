import { registerAs } from '@nestjs/config';

export const endpointsConfig = registerAs('endpoints', () => ({
  status: {
    url: '/status',
    method: 'GET',
  },
  accesstoken: {
    url: '/accesstoken',
    method: 'GET',
  },
  accesstokenrefresh: {
    url: '/accesstoken?refresh=true',
    method: 'GET',
  },
}));
