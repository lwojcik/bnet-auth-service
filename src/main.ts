import { NestFactory } from '@nestjs/core';
import {
  FastifyAdapter,
  NestFastifyApplication,
} from '@nestjs/platform-fastify';
import fs from 'fs';
import { ValidationPipe } from '@nestjs/common';
import { SwaggerModule, DocumentBuilder } from '@nestjs/swagger';
import { ConfigService } from '@nestjs/config';
import { AppModule } from './app.module';
import { APP, APP_INFO, HTTPS } from './common/constants';
import { Environment } from './common/types';
import { trueStringToBoolean } from './utils/trueStringToBoolean';

async function bootstrap() {
  const app = await NestFactory.create<NestFastifyApplication>(
    AppModule,
    new FastifyAdapter(
      trueStringToBoolean({ value: process.env[HTTPS.enable] })
        ? {
            https: {
              key: fs.readFileSync(process.env[HTTPS.keyPath]),
              cert: fs.readFileSync(process.env[HTTPS.certPath]),
            },
          }
        : undefined
    )
  );
  const configService = app.get(ConfigService);
  const port = configService.get<string>(APP.port);
  const host = configService.get<string>(APP.host);

  if (trueStringToBoolean({ value: process.env[APP.enableCors] })) {
    const corsConfig = process.env[APP.corsOrigin]
      ? {
          origin: process.env[APP.corsOrigin],
          methods: ['GET'],
        }
      : undefined;
    app.enableCors(corsConfig);
  }

  if (process.env[APP.environment] !== Environment.production) {
    const swaggerConfig = new DocumentBuilder()
      .setTitle(APP_INFO.name)
      .setDescription(APP_INFO.description)
      .setVersion(APP_INFO.version)
      .build();
    const document = SwaggerModule.createDocument(app, swaggerConfig);
    SwaggerModule.setup('api', app, document);
  }

  app.useGlobalPipes(
    new ValidationPipe({
      transform: true,
    })
  );

  await app.listen(port, host);
}
bootstrap();
