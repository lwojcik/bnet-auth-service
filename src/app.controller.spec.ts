import { ConfigModule, registerAs } from '@nestjs/config';
import { Test, TestingModule } from '@nestjs/testing';
import { AppController } from './app.controller';
import { LoggerService } from './logger/logger.service';
import { MainService } from './main/main.service';

describe('AppController', () => {
  let appController: AppController;

  beforeEach(async () => {
    const app: TestingModule = await Test.createTestingModule({
      imports: [
        ConfigModule.forFeature(
          registerAs('endpoints', () => ({
            testEndpoint1: {
              name: 'testEndpoint1',
              url: '/testEndpoint1',
              method: 'TEST_METHOD_1',
            },
            testEndpoint2: {
              name: 'testEndpoint2',
              url: '/testEndpoint2',
              method: 'TEST_METHOD_2',
            },
          }))
        ),
        ConfigModule.forFeature(
          registerAs('redis', () => ({
            enable: false,
          }))
        ),
      ],
      controllers: [AppController],
      providers: [
        MainService,
        {
          provide: LoggerService,
          useValue: {
            debug: jest.fn(),
            log: jest.fn(),
            setLoggedClass: jest.fn(),
            setLoggedMethod: jest.fn(),
          },
        },
      ],
    }).compile();

    appController = app.get<AppController>(AppController);
  });

  it('should implement getMain method', () => {
    expect(() => {
      appController.getMain();
    }).not.toThrow();
  });
});
