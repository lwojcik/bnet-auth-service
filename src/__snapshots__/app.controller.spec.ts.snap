// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppController should return name and list of routes 1`] = `
Object {
  "caching": true,
  "endpoints": Object {
    "accesstoken": Object {
      "method": "GET",
      "url": "/accesstoken",
    },
    "status": Object {
      "method": "GET",
      "url": "/status",
    },
  },
  "name": "bnet-auth-service",
  "version": "2.0.0",
}
`;
